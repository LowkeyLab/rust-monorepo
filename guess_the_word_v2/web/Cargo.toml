[package]
name = "guess_the_word_v2"
version = "0.1.0"
authors = ["tacascer <trandangthanh2000@gmail.com>"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
dioxus = { version = "0.6.0", features = ["router", "fullstack"] }
guess_the_word_v2_core = { path = "../core", features = ["serde"] }
serde = { version = "1.0", features = ["derive"]}

[features]
default = []
# The feature that are only required for the web build target should be optional and only enabled in the web feature
web = ["dioxus/web"]
# The feature that are only required for the desktop build target should be optional and only enabled in the desktop feature
desktop = ["dioxus/desktop"]
# The feature that are only required for the mobile build target should be optional and only enabled in the mobile feature
mobile = ["dioxus/mobile"]
# The feature that are only required for the server build target should be optional and only enabled in the server feature
server = ["dioxus/server"]

