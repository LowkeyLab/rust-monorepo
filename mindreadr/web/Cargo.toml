[package]
name = "mindreadr_web"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
dioxus = { version = "0.6", features = ["router", "fullstack"] }
mindreadr_core = { path = "../core", features = ["serde"] }
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
tokio = { version = "1.47.1", features = ["full"], optional = true }
axum = { version = "0.7", optional = true }
mindreadr_web_migration = { path = "./migration", optional = true }
sea-orm = { version = "1.1.0", features = ["runtime-tokio-rustls", "sqlx-postgres"], optional = true }
tracing = "0.1.41"
tracing-futures = "0.2.5"
config = { version = "0.15.14", optional = true }
anyhow = "1.0.99"
thiserror = "2.0.16"
gloo-storage = "0.3"

[features]
default = ["web"]
# The feature that are only required for the web build target should be optional and only enabled in the web feature
web = ["dioxus/web"]
# The feature that are only required for the desktop build target should be optional and only enabled in the desktop feature
desktop = ["dioxus/desktop"]
# The feature that are only required for the mobile build target should be optional and only enabled in the mobile feature
mobile = ["dioxus/mobile"]
# The feature that are only required for the server build target should be optional and only enabled in the server feature
server = ["dioxus/server", "tokio", "axum", "mindreadr_web_migration", "sea-orm", "config"]
