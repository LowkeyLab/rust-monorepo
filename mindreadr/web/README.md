# Development

## Tailwind
Run the following command:

```bash
mise run tailwind
```

## Serving Your App

Run the following command in the root of your project to start developing with the default platform:

```bash
mise run-local
```

To run for a different platform, use the `--platform platform` flag. E.g.
```bash
dx serve --platform desktop
```

