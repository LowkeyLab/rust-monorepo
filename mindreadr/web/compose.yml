services:
  postgres:
    image: postgres:latest
    container_name: postgres_db
    restart: unless-stopped
    environment:
      POSTGRES_USER: postgres
      POSTGRES_PASSWORD: password
      POSTGRES_DB: guess_the_word
    ports:
      - "5432:5432"
    volumes:
      - postgres_data:/var/lib/postgresql/data
      - ./init.sql:/docker-entrypoint-initdb.d/init.sql
    networks:
      - app_network

  pgadmin:
    image: dpage/pgadmin4:latest
    container_name: pgadmin
    restart: unless-stopped
    environment:
      PGADMIN_DEFAULT_EMAIL: admin@example.com
      PGADMIN_DEFAULT_PASSWORD: admin
    ports:
      - "5000:80"
    volumes:
      - pgadmin_data:/var/lib/pgadmin
    networks:
      - app_network
    depends_on:
      - postgres

volumes:
  postgres_data:
  pgadmin_data:

networks:
  app_network:
    driver: bridge
