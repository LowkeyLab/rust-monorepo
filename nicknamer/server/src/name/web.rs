use askama::Template;
use axum::{
    Form, Router,
    extract::{RawQuery, State},
    http::{HeaderMap, HeaderName, HeaderValue, StatusCode},
    response::Html,
    routing::get,
};
use serde::Deserialize;
use std::sync::Arc;

use crate::name::{Name, NameService, NameServiceError};

#[derive(Debug, Deserialize)]
pub struct CreateNameForm {
    discord_id: u64,
    name: String,
    server_id: String,
}

#[derive(Debug, Deserialize)]
pub struct EditNameForm {
    name: String,
    server_id: String,
}

#[derive(Debug, Deserialize)]
pub struct BulkAddForm {
    server_id: String,
    yaml_content: String,
}

/// Helper function to get all names, sort them using the provided function, and render them as a names table.
/// This reduces code duplication across handlers that need to display sorted names.
#[tracing::instrument(skip(name_service, sort_fn))]
async fn render_names_table<F>(
    name_service: &NameService<'_>,
    sort_fn: F,
) -> Result<String, NameError>
where
    F: FnOnce(&mut Vec<Name>),
{
    let mut names = name_service.get_all_names().await?;
    sort_fn(&mut names);
    let table_template = NamesTableTemplate::new(names);
    table_template.render().map_err(NameError::from)
}

/// Custom error type for name handler operations.
#[derive(Debug, thiserror::Error)]
enum NameError {
    /// Represents an error during template rendering.
    #[error("Template rendering failed")]
    Template(#[from] askama::Error),
    /// Represents a database error.
    #[error("Database error")]
    Database(#[from] anyhow::Error),
    /// Represents a name service error.
    #[error("Name service error")]
    Service(#[from] NameServiceError),
    /// Represents a duplicate entry error (Discord ID + Server ID combination already exists).
    #[error("A name entry already exists for this Discord ID and Server ID combination")]
    DuplicateEntry,
    /// Represents an I/O error.
    #[error("I/O error: {0}")]
    Io(#[from] std::io::Error),
}

impl axum::response::IntoResponse for NameError {
    fn into_response(self) -> axum::response::Response {
        let (status_code, user_facing_error_message) = match self {
            NameError::DuplicateEntry => (
                StatusCode::UNPROCESSABLE_ENTITY,
                "A name entry already exists for this Discord ID and Server ID combination. Please use a different combination.",
            ),
            _ => (
                StatusCode::INTERNAL_SERVER_ERROR,
                "An unexpected error occurred while processing your request. Please try again later.",
            ),
        };

        let error_template = ErrorMessageTemplate::new(user_facing_error_message.to_string());
        let Ok(rendered) = error_template.render() else {
            return StatusCode::INTERNAL_SERVER_ERROR.into_response();
        };

        let mut response = (status_code, Html(rendered)).into_response();
        // Add HTMX headers to retarget the error message to the error div
        let mut headers = HeaderMap::new();
        headers.insert(
            HeaderName::from_static("hx-reswap"),
            HeaderValue::from_static("innerHTML"),
        );
        response.headers_mut().extend(headers);
        response
    }
}

#[derive(Template)]
#[template(path = "names.html")]
struct NamesTemplate {}

impl NamesTemplate {
    pub fn new() -> Self {
        Self {}
    }
}

#[derive(Template)]
#[template(path = "names/add_name_form.html")]
struct AddNameFormTemplate;

#[derive(Template)]
#[template(path = "names/names_table.html")]
struct NamesTableTemplate {
    names: Vec<Name>,
}

impl NamesTableTemplate {
    pub fn new(names: Vec<Name>) -> Self {
        Self { names }
    }
}

#[derive(Template)]
#[template(path = "names/error_message.html")]
struct ErrorMessageTemplate {
    message: String,
}

impl ErrorMessageTemplate {
    pub fn new(message: String) -> Self {
        Self { message }
    }
}

#[derive(Template)]
#[template(path = "names/edit_name_form.html")]
struct EditNameFormTemplate {
    name: Name,
}

impl EditNameFormTemplate {
    pub fn new(name: Name) -> Self {
        Self { name }
    }
}

#[derive(Template)]
#[template(path = "names/name_row.html")]
struct NameRowTemplate {
    name: Name,
}

impl NameRowTemplate {
    pub fn new(name: Name) -> Self {
        Self { name }
    }
}

#[derive(Template)]
#[template(path = "names/bulk_add_form.html")]
struct BulkAddFormTemplate;

#[derive(Template)]
#[template(path = "names/bulk_add_success.html")]
struct BulkAddSuccessTemplate {
    created_count: usize,
    skipped_count: usize,
    errors: Vec<String>,
    server_id: String,
    yaml_content: String,
}

impl BulkAddSuccessTemplate {
    pub fn new(
        created_count: usize,
        skipped_count: usize,
        errors: Vec<String>,
        server_id: String,
        yaml_content: String,
    ) -> Self {
        Self {
            created_count,
            skipped_count,
            errors,
            server_id,
            yaml_content,
        }
    }
}

#[derive(Template)]
#[template(path = "names/bulk_delete.html")]
struct BulkDeleteTemplate;

#[derive(Template)]
#[template(path = "names/bulk_delete_table.html")]
struct BulkDeleteTableTemplate {
    names: Vec<Name>,
}

impl BulkDeleteTableTemplate {
    pub fn new(names: Vec<Name>) -> Self {
        Self { names }
    }
}

#[derive(Clone, Debug)]
pub struct NameState {
    pub db: Arc<sea_orm::DatabaseConnection>,
}

/// Handler for the /names endpoint that displays all names in a table.
#[tracing::instrument]
async fn names_handler() -> Result<Html<String>, NameError> {
    let template = NamesTemplate::new();
    template.render().map(Html).map_err(NameError::from)
}

/// Handler for creating a new name via POST request.
#[tracing::instrument(skip(state))]
async fn create_name_handler(
    State(state): State<Arc<NameState>>,
    Form(form): Form<CreateNameForm>,
) -> Result<Html<String>, NameError> {
    let name_service = NameService::new(&state.db);

    match name_service
        .create_name(form.discord_id, form.name, form.server_id)
        .await
    {
        Ok(_) => {
            // Get updated names for the table and render
            let table_html = render_names_table(&name_service, |names| {
                names.sort_by_key(|name| name.id());
            })
            .await?;
            Ok(Html(table_html))
        }
        Err(NameServiceError::DuplicateEntryError(_, _)) => Err(NameError::DuplicateEntry),
        Err(err) => Err(NameError::Service(err)),
    }
}

/// Handler for serving the add name form.
#[tracing::instrument]
async fn add_name_form_handler() -> Result<Html<String>, NameError> {
    let template = AddNameFormTemplate;
    template.render().map(Html).map_err(NameError::from)
}

/// Handler for deleting a name via POST request.
#[tracing::instrument(skip(state))]
async fn delete_name_handler(
    State(state): State<Arc<NameState>>,
    axum::extract::Path(id): axum::extract::Path<u32>,
) -> Result<Html<String>, NameError> {
    let name_service = NameService::new(&state.db);

    match name_service.delete_name_by_id(id).await {
        Ok(_) => {
            // Get updated names for the table and render
            let table_html = render_names_table(&name_service, |names| {
                names.sort_by_key(|name| name.id());
            })
            .await?;
            Ok(Html(table_html))
        }
        Err(err) => Err(NameError::Service(err)),
    }
}

/// Handler for bulk deleting names via DELETE request.
#[tracing::instrument(skip(state))]
async fn bulk_delete_names_handler(
    State(state): State<Arc<NameState>>,
    RawQuery(query): RawQuery,
) -> Result<Html<String>, NameError> {
    let name_service = NameService::new(&state.db);

    // Parse query parameters manually to handle multiple values with the same key
    let selected_ids: Vec<u32> = if let Some(query_str) = query {
        query_str
            .split('&')
            .filter_map(|pair| {
                if pair.starts_with("selected_ids=") {
                    pair.strip_prefix("selected_ids=")
                        .and_then(|id_str| id_str.parse().ok())
                } else {
                    None
                }
            })
            .collect()
    } else {
        Vec::new()
    };

    if selected_ids.is_empty() {
        // No names selected for deletion, just return the current table
        let table_html = render_names_table(&name_service, |names| {
            names.sort_by_key(|name| name.id());
        })
        .await?;
        return Ok(Html(table_html));
    }

    match name_service.bulk_delete_names(&selected_ids).await {
        Ok(_) => {
            // Get updated names for the table and render
            let table_html = render_names_table(&name_service, |names| {
                names.sort_by_key(|name| name.id());
            })
            .await?;
            Ok(Html(table_html))
        }
        Err(err) => Err(NameError::Service(err)),
    }
}

/// Handler for serving the edit name form.
#[tracing::instrument(skip(state))]
async fn edit_name_handler(
    State(state): State<Arc<NameState>>,
    axum::extract::Path(id): axum::extract::Path<u32>,
) -> Result<Html<String>, NameError> {
    let name_service = NameService::new(&state.db);

    match name_service.get_name_by_id(id).await {
        Ok(name) => {
            let template = EditNameFormTemplate::new(name);
            template.render().map(Html).map_err(NameError::from)
        }
        Err(err) => Err(NameError::Service(err)),
    }
}

/// Handler for updating a name via PUT request.
#[tracing::instrument(skip(state))]
async fn update_name_handler(
    State(state): State<Arc<NameState>>,
    axum::extract::Path(id): axum::extract::Path<u32>,
    Form(form): Form<EditNameForm>,
) -> Result<Html<String>, NameError> {
    let name_service = NameService::new(&state.db);

    match name_service
        .edit_name_by_id(id, form.name, form.server_id)
        .await
    {
        Ok(_) => {
            // Get the updated name to render just this row
            let updated_name = name_service.get_name_by_id(id).await?;

            // Render only the updated name row
            let row_template = NameRowTemplate::new(updated_name);
            let row_html = row_template.render().map_err(NameError::from)?;

            Ok(Html(row_html))
        }
        Err(err) => Err(NameError::Service(err)),
    }
}

/// Handler for GET /names/table that returns just the names table fragment.
#[tracing::instrument(skip(state))]
async fn names_table_handler(
    State(state): State<Arc<NameState>>,
) -> Result<Html<String>, NameError> {
    let name_service = NameService::new(&state.db);
    let table_html = render_names_table(&name_service, |names| {
        names.sort_by_key(|name| name.id());
    })
    .await?;
    Ok(Html(table_html))
}

/// Handler for GET /names/{id} that returns a single name row.
#[tracing::instrument(skip(state))]
async fn get_name_row_handler(
    State(state): State<Arc<NameState>>,
    axum::extract::Path(id): axum::extract::Path<u32>,
) -> Result<Html<String>, NameError> {
    let name_service = NameService::new(&state.db);

    match name_service.get_name_by_id(id).await {
        Ok(name) => {
            let template = NameRowTemplate::new(name);
            template.render().map(Html).map_err(NameError::from)
        }
        Err(err) => Err(NameError::Service(err)),
    }
}

/// Handler for GET /names/bulk-add that displays the bulk add form.
#[tracing::instrument]
async fn bulk_add_form_handler() -> Result<Html<String>, NameError> {
    let template = BulkAddFormTemplate;
    template.render().map(Html).map_err(NameError::from)
}

/// Handler for POST /names/bulk-add that processes the YAML content.
#[tracing::instrument(skip(state, form))]
async fn bulk_add_handler(
    State(state): State<Arc<NameState>>,
    Form(form): Form<BulkAddForm>,
) -> Result<Html<String>, NameError> {
    let name_service = NameService::new(&state.db);

    // Process the bulk upload using the pasted YAML content
    match name_service
        .bulk_create_names(&form.yaml_content, form.server_id.clone())
        .await
    {
        Ok((created_count, skipped_count, errors)) => {
            let template = BulkAddSuccessTemplate::new(
                created_count,
                skipped_count,
                errors,
                form.server_id,
                form.yaml_content,
            );
            template.render().map(Html).map_err(NameError::from)
        }
        Err(err) => Err(NameError::Service(err)),
    }
}

/// Handler for GET /names/delete that displays the bulk delete interface.
#[tracing::instrument]
async fn bulk_delete_page_handler() -> Result<Html<String>, NameError> {
    let template = BulkDeleteTemplate;
    template.render().map(Html).map_err(NameError::from)
}

/// Handler for GET /names/delete/table that returns the bulk delete table fragment.
#[tracing::instrument(skip(state))]
async fn bulk_delete_table_handler(
    State(state): State<Arc<NameState>>,
) -> Result<Html<String>, NameError> {
    let name_service = NameService::new(&state.db);
    let mut names = name_service.get_all_names().await?;
    names.sort_by_key(|name| name.id());
    let template = BulkDeleteTableTemplate::new(names);
    template.render().map(Html).map_err(NameError::from)
}

/// Handler for DELETE /names/delete that processes bulk deletion and returns the updated bulk delete table.
#[tracing::instrument(skip(state))]
async fn bulk_delete_names_delete_handler(
    State(state): State<Arc<NameState>>,
    RawQuery(query): RawQuery,
) -> Result<Html<String>, NameError> {
    let name_service = NameService::new(&state.db);

    // Parse query parameters manually to handle multiple values with the same key
    let selected_ids: Vec<u32> = if let Some(query_str) = query {
        query_str
            .split('&')
            .filter_map(|pair| {
                if pair.starts_with("selected_ids=") {
                    pair.strip_prefix("selected_ids=")
                        .and_then(|id_str| id_str.parse().ok())
                } else {
                    None
                }
            })
            .collect()
    } else {
        Vec::new()
    };

    // Perform the deletion if any IDs are selected
    if !selected_ids.is_empty() {
        let _ = name_service.bulk_delete_names(&selected_ids).await;
    }

    // Return the updated bulk delete table (same as GET /names/delete/table)
    let mut names = name_service.get_all_names().await?;
    names.sort_by_key(|name| name.id());
    let template = BulkDeleteTableTemplate::new(names);
    template.render().map(Html).map_err(NameError::from)
}

/// Creates and returns the name router with all name-related routes.
pub fn create_name_router(state: Arc<NameState>) -> Router {
    Router::new()
        .route(
            "/names",
            get(names_handler)
                .post(create_name_handler)
                .delete(bulk_delete_names_handler),
        )
        .route("/names/add", get(add_name_form_handler))
        .route(
            "/names/bulk-add",
            get(bulk_add_form_handler).post(bulk_add_handler),
        )
        .route(
            "/names/delete",
            get(bulk_delete_page_handler).delete(bulk_delete_names_delete_handler),
        )
        .route("/names/delete/table", get(bulk_delete_table_handler))
        .route(
            "/names/{id}",
            get(get_name_row_handler)
                .delete(delete_name_handler)
                .put(update_name_handler),
        )
        .route("/names/{id}/edit", get(edit_name_handler))
        .route("/names/table", get(names_table_handler))
        .with_state(state.clone())
}
