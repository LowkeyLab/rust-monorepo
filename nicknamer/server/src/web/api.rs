pub(crate) mod v1 {
    use std::sync::Arc;

    use crate::{
        auth::{self, AuthState},
        name::NameState,
    };

    use axum::{
        Router,
        middleware::{from_fn, from_fn_with_state},
    };

    use tower::ServiceBuilder;
    use utoipa::OpenApi;
    use utoipa_swagger_ui::SwaggerUi;

    /// OpenAPI documentation for the v1 API
    #[derive(OpenApi)]
    #[openapi(
        paths(
            crate::auth::api::v1::json_login_handler,
            crate::name::api::v1::get_names_handler,
        ),
        components(
            schemas(
                crate::auth::api::v1::JsonLoginRequest,
                crate::auth::api::v1::LoginResponse,
                crate::auth::api::v1::ErrorResponse,
                crate::name::api::v1::NameJson,
                crate::name::api::v1::NamesResponse,
                crate::name::api::v1::ErrorResponse,
            )
        ),
        tags(
            (name = "Authentication", description = "Authentication endpoints"),
            (name = "Names", description = "Name management endpoints")
        ),
        info(
            title = "Nicknamer API",
            version = "1.0.0",
            description = "API for managing Discord nicknames and authentication"
        )
    )]
    pub struct ApiDoc;

    /// Creates the API routes for JSON API endpoints.
    pub fn create_api_router(
        auth_state: Arc<AuthState>,
        name_state: Arc<NameState>,
    ) -> axum::Router {
        let login_router = auth::api::v1::create_api_router(auth_state.clone());
        let names_router = crate::name::api::v1::create_api_router(name_state.clone());
        let protected_routes = names_router
            .layer(ServiceBuilder::new().layer(from_fn(auth::api::v1::require_auth_middleware)));
        let public_routes = login_router;
        let api_routes = public_routes.merge(protected_routes);

        Router::new()
            .merge(SwaggerUi::new("/swagger-ui").url("/api-docs/openapi.json", ApiDoc::openapi()))
            .nest("/api/v1", api_routes)
            .layer(ServiceBuilder::new().layer(from_fn_with_state(
                auth_state,
                auth::api::v1::auth_user_middleware,
            )))
    }
}
