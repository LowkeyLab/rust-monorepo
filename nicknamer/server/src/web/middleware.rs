use axum::extract::Request;
use axum::http::{HeaderName, HeaderValue};
use axum::middleware::Next;
use axum::response::Response;

/// Middleware function that adds CORS headers to expose HTMX headers
pub async fn cors_expose_headers(request: Request, next: Next) -> Response {
    let mut response = next.run(request).await;

    // Add the Access-Control-Expose-Headers header
    response.headers_mut().insert(
        HeaderName::from_static("access-control-expose-headers"),
        HeaderValue::from_static("hx-retarget,hx-reswap"),
    );

    response
}

#[cfg(test)]
mod tests {
    use super::*;
    use axum::body::Body;
    use axum::http::{Request, StatusCode};
    use axum::{Router, response::Response};
    use tower::ServiceExt;

    #[tokio::test]
    async fn can_add_cors_expose_header() {
        let app = Router::new()
            .route("/test", axum::routing::get(|| async { "test response" }))
            .layer(axum::middleware::from_fn(cors_expose_headers));

        let response = app
            .oneshot(
                Request::builder()
                    .method("GET")
                    .uri("/test")
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);

        let headers = response.headers();
        let expose_headers = headers.get("access-control-expose-headers");
        assert_eq!(
            expose_headers,
            Some(&axum::http::HeaderValue::from_static(
                "hx-retarget,hx-reswap"
            ))
        );
    }

    #[tokio::test]
    async fn can_preserve_existing_headers_when_adding_cors() {
        async fn handler_with_custom_header() -> Response<String> {
            let mut response = Response::new("test response".to_string());
            response.headers_mut().insert(
                "custom-header",
                axum::http::HeaderValue::from_static("custom-value"),
            );
            response
        }

        let app = Router::new()
            .route(
                "/test-with-headers",
                axum::routing::get(handler_with_custom_header),
            )
            .layer(axum::middleware::from_fn(cors_expose_headers));

        let response = app
            .oneshot(
                Request::builder()
                    .method("GET")
                    .uri("/test-with-headers")
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);

        let headers = response.headers();

        // Check that our CORS header was added
        let expose_headers = headers.get("access-control-expose-headers");
        assert_eq!(
            expose_headers,
            Some(&axum::http::HeaderValue::from_static(
                "hx-retarget,hx-reswap"
            ))
        );

        // Check that existing headers are preserved
        let custom_header = headers.get("custom-header");
        assert_eq!(
            custom_header,
            Some(&axum::http::HeaderValue::from_static("custom-value"))
        );
    }
}
