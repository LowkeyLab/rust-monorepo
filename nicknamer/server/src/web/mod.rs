use askama::Template;
use axum::http::StatusCode;
use axum::response::Html;
use migration::MigratorTrait;
use sea_orm::Database;
use std::sync::Arc;
use tower_http::trace::{DefaultMakeSpan, DefaultOnRequest, DefaultOnResponse, TraceLayer};
use tracing::Level;

use crate::auth::{create_login_router, AuthState};
use crate::config::{self, Config};

pub mod middleware;

#[derive(Clone)]
pub struct AppState {
    pub config: Arc<Config>,
}

/// Custom error type for web handler operations.
#[derive(Debug, thiserror::Error)]
enum WebError {
    /// Represents an error during template rendering.
    /// The specific `askama::Error` is captured as the source of this error.
    #[error("Template rendering failed")]
    Template(#[from] askama::Error),
}

impl axum::response::IntoResponse for WebError {
    fn into_response(self) -> axum::response::Response {
        let user_facing_error_message =
            "An unexpected error occurred while processing your request. Please try again later.";
        (
            StatusCode::INTERNAL_SERVER_ERROR,
            Html(format!(
                "<h1>Internal Server Error</h1><p>{}</p>",
                user_facing_error_message
            )),
        )
            .into_response()
    }
}

#[tracing::instrument]
pub async fn start_web_server(config: config::Config) -> anyhow::Result<()> {
    use axum::Router;

    let server_address = format!("0.0.0.0:{}", &config.port);
    let listener = tokio::net::TcpListener::bind(&server_address).await?;
    tracing::info!("Web server running on http://{}", server_address);

    let db = Database::connect(&config.db_url).await?;
    migration::Migrator::up(&db, None).await?;
    tracing::info!("Database migrations applied successfully");

    // Create AuthState from config
    let auth_state = Arc::new(AuthState::from_config(&config));

    // Create the login router with AuthState
    let login_router = create_login_router(auth_state.clone());

    let main_router = Router::new()
        .route("/health", axum::routing::get(health_check_handler))
        .route("/", axum::routing::get(welcome_handler))
        .layer(
            TraceLayer::new_for_http()
        )
        ;

    // Create main router and merge with login router
    let app = Router::new().merge(main_router).merge(login_router);

    axum::serve(listener, app).await?;
    Ok(())
}

#[tracing::instrument]
async fn health_check_handler() -> &'static str {
    "OK"
}

#[tracing::instrument]
async fn welcome_handler() -> Result<Html<String>, WebError> {
    IndexTemplate.render().map(Html).map_err(WebError::from)
}

#[derive(Template)]
#[template(path = "index.html")]
struct IndexTemplate;

#[cfg(test)]
mod tests {
    use super::*;
    use axum::http::StatusCode;

    #[tokio::test]
    async fn can_check_health() {
        let response = health_check_handler().await;
        assert_eq!(response, "OK");
    }

    #[tokio::test]
    async fn can_render_welcome_page_with_correct_content_type() {
        let result = welcome_handler().await;
        assert!(
            result.is_ok(),
            "welcome() returned an error: {:?}",
            result.err()
        );
        let response: axum::response::Response =
            axum::response::IntoResponse::into_response(result.unwrap());
        assert_eq!(
            response.status(),
            StatusCode::OK,
            "Welcome handler should return OK for this test"
        );
        let content_type = response.headers().get(axum::http::header::CONTENT_TYPE);
        assert_eq!(
            content_type,
            Some(&axum::http::HeaderValue::from_static(
                "text/html; charset=utf-8"
            ))
        );
    }

    #[tokio::test]
    async fn can_handle_template_error_with_internal_server_error() {
        // Simulate a template rendering error using askama::Error::Custom
        let custom_error_message = "Simulated template rendering failure".to_string();
        let template_error = askama::Error::Custom(custom_error_message.into());

        let web_error = WebError::Template(template_error);
        let response = axum::response::IntoResponse::into_response(web_error);

        assert_eq!(response.status(), StatusCode::INTERNAL_SERVER_ERROR);

        let body = axum::body::to_bytes(response.into_body(), usize::MAX)
            .await
            .unwrap();
        let expected_error_message = "<h1>Internal Server Error</h1><p>An unexpected error occurred while processing your request. Please try again later.</p>";
        assert_eq!(std::str::from_utf8(&body).unwrap(), expected_error_message);
    }
}
