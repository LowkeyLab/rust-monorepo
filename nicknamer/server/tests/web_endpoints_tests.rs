use axum::body::Body;
use axum::http::Request;
use axum::Router;
use insta::assert_yaml_snapshot;
use nicknamer_server::web::{health_check_handler, welcome_handler, call_to_action_handler};
use tower::ServiceExt;

mod common;

use common::HttpResponseSnapshot;

/// Create a router for testing web endpoints.
/// This function creates a minimal router with just the public routes needed for testing.
fn create_test_router() -> Router {
    Router::new()
        .route("/health", axum::routing::get(health_check_handler))
        .route("/", axum::routing::get(welcome_handler))
        .route(
            "/call-to-action",
            axum::routing::get(call_to_action_handler),
        )
}

#[tokio::test]
async fn can_render_welcome_page() {
    let app = create_test_router();
    
    let request = Request::builder()
        .uri("/")
        .body(Body::empty())
        .unwrap();

    let response = app.oneshot(request).await.unwrap();
    
    let status = response.status();
    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot = HttpResponseSnapshot::new(body_text, status, &headers, "welcome_page");
    assert_yaml_snapshot!(snapshot);
}

#[tokio::test]
async fn can_render_call_to_action_for_unauthenticated_user() {
    let app = create_test_router();
    
    let request = Request::builder()
        .uri("/call-to-action")
        .body(Body::empty())
        .unwrap();

    let response = app.oneshot(request).await.unwrap();
    
    let status = response.status();
    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot = HttpResponseSnapshot::new(body_text, status, &headers, "call_to_action_unauthenticated");
    assert_yaml_snapshot!(snapshot);
}

#[tokio::test]
async fn can_check_health_endpoint() {
    let app = create_test_router();
    
    let request = Request::builder()
        .uri("/health")
        .body(Body::empty())
        .unwrap();

    let response = app.oneshot(request).await.unwrap();
    
    let status = response.status();
    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot = HttpResponseSnapshot::new(body_text, status, &headers, "health_check");
    assert_yaml_snapshot!(snapshot);
}
