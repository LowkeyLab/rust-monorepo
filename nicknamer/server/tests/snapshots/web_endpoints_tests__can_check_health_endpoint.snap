---
source: nicknamer/server/tests/web_endpoints_tests.rs
expression: snapshot
---
test_context: health_check
status: 200
headers:
  content-type: text/plain; charset=utf-8
html_body:
  - OK
